### ISSUE TITLE
Install Service Fabric SDK as pipeline task

### ISSUE THREAD

**Issue by jackbond** [microsoft/service-fabric#333]
I updated to the latest Service Fabric SDK, and updated the NuGet packages, and version numbers in the project files, and everything built great locally. When I checked my changes in, it broke our build pipeline. For the standard "Hosted Windows 2019 with VS2019" it shows the following SF info...

Azure Service Fabric
SDK Version: 3.3.617.9590<br/>
Runtime Version: 6.4.617.9590

So, how does one install the Service Fabric SDK as part of a devops pipeline. Thanks.

**Comment by jackbond** [comment]
Common guys, this is bullshit. It's completely unacceptable that the SF team doesn't have a person whose job it is to monitor issues raised here.

There is a "Service Fabric Application Deployment" task officially support in DevOps pipelines, which is COMPLETELY EFFING USELESS, if you can't use the latest SDK. Glaring gaps like this make Microsoft look amateurish.

**Comment by juho-hanhimaki** [comment]
Yep the experience is not great. Somewhere in the internet there is a document that states the versions of software that is installed on the devops image. The devops UI should offer clarity for versioning so one doesn't have to hunt for the documentation after having encountered errors.

Don't know anything about installing the latest SDK manually. Would be great if there was option to do that. Maybe there is?

**Comment by jackbond** [comment]
My powershell script actually fails on the uninstall. Bottom line, there's a .Netcore sdk installer task, a NuGet installer task, why isn't there a Service Fabric SDK installer task? If not that, have a clear set of instructions for upgrading SF via powershell. Finally, it's slightly out of their hands, but why is the build server's SDK stuck at 3.3.617.9590?

**Comment by masnider** [comment]
The VSTS issues are best addressed in the VSTS tasks repo, which is over [here]( https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks). When you tried with the instructions in our docs, did you run into a problem? When you say it "broke your build pipeline" can you be more specific? Do the instructions [here](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-tutorial-deploy-app-with-cicd-vsts) not work?

**Comment by masnider** [comment]
@jackbond Hey Jack. If you're going to light up the team, we would appreciate a response and more information so that we can help.

**Comment by jackbond** [comment]
@masnider I'm in the process of reproducing the issue.

**Comment by jackbond** [comment]
I have at least 10 different service fabric solutions all building and deploying via Azure DevOps. I never had an issue until about a month ago when I updated to the latest SF NuGet packages including Microsoft.VisualStudio.Azure.Fabric.MSBuild (from 1.6.7. to 1.6.8). From the image below you can see that when I updated all the SF packages, the pipeline failed to build (allow the code built locally). I then updated all the SF NuGets except Microsoft.VisualStudio.Azure.Fabric.MSBuild, and the pipeline succeeded. So there appears to be an issue between Microsoft.VisualStudio.Azure.Fabric.MSBuild 1.6.8 and the SDK installed on the "Hosted Windows 2019 with VS2019" agent pool.

![image](https://user-images.githubusercontent.com/2799232/59712022-295a2500-91c1-11e9-8312-00a31676ef29.png)

**Comment by bishal-pdMSFT** [comment]
@jackbond the Azure Pipelines hosted images get updated once a month and hence it can take upto a month for newly publicly released SDK version to show up on Hosted agents. By publicly I mean that the new version should be available on Web platform installer.
If you want to use hosted agents, you can think about reverting to older MSBuild package. Is there a critical fix you want to take from newer package? In that case, your best bet would be to use a Private agent (with latest version of SDK installed)

**Comment by bishal-pdMSFT** [comment]
@jackbond can you please attach your build logs for us to check if this is really an issue with SDK version. You can generate debug logs by setting a build variable called _system.debug_ with value _true_.

**Comment by jackbond** [comment]
Is there a non-public email I can send them to?

**Comment by bishal-pdMSFT** [comment]
You can send them to RM_Customer_Queries @ microsoft.com

**Comment by jackbond** [comment]
I have the Powershell script below in a pipeline, and have attached the associated log. It's failing with:

MSI (s) (B8:88) [18:22:49:094]: Product: Microsoft Azure Service Fabric SDK -- Error 1714. The older version of Microsoft Azure Service Fabric SDK cannot be removed.  Contact your technical support group.  System Error 1612.

Any ideas?

[log_5_3661.zip](https://github.com/microsoft/service-fabric/files/3331516/log_5_3661.zip)


```
#Uninstall existing SDK

$url = "https://download.microsoft.com/download/C/7/B/C7B7BA41-B95A-4363-8B28-7075E9B10E99/MicrosoftServiceFabricSDK.3.4.639.msi"

$output = "MicrosoftServiceFabricSDK.3.4.639.msi"

Invoke-WebRequest -Uri $url -OutFile $output

Write-Host "SDK Transfer Complete"

Start-Process -Wait -FilePath msiexec.exe -ArgumentList "/i MicrosoftServiceFabricSDK.3.4.639.msi /quiet /l*v SFSdkInstall.log"

Get-Content SFSdkInstall.log | foreach {Write-Output $_}

Get-CimInstance -Query "Select * from Win32_Product where Name like 'Microsoft Azure Service%'"
```

**Comment by bishal-pdMSFT** [comment]
@athinanthny @masnider I think this question is for you.

**Comment by masnider** [comment]
@jackbond I would open a separate issue for this so that we don't lose track of your original issue.

**Comment by ravipal** [comment]
1.6.8 version of Microsoft.VisualStudio.Azure.Fabric.MSBuild have an optimization change that avoids building the service project second time during package and causes this issue. The default core project created by VS will use platform=‘Any CPU’ during solution build. So during Package, the binaries are available when the service project’s package task (DotNetPublish) runs. From the attached log, the solution build is using the X64 platform and package is using the ‘Any CPU’, hence the package fails. Earlier version of nuget package was doing an additional build during package which uses the same platform as package, so it was working. My suggestion is to update the solution configuration to always build ‘Any CPU’ for service projects, similar to the default project created by VS.

**Comment by bishal-pdMSFT** [comment]
@ravipal thank for looking into this. Just to confirm that I understand your terminology - the core project you are referring to is .sfproj. Is that correct?
In Azure DevOps pipeline, by default we create a pipeline which has x64 platform for both solution build as well as sfproj build. Is this combination going to work?
![image](https://user-images.githubusercontent.com/7830141/60940938-e8a28700-a2fa-11e9-9e57-c5839779d8df.png)
![image](https://user-images.githubusercontent.com/7830141/60940953-f9eb9380-a2fa-11e9-9dd1-c27985c876f0.png)
![image](https://user-images.githubusercontent.com/7830141/60940971-0cfe6380-a2fb-11e9-95e8-abdeb2321e0f.png)

@jackbond can you please check if you have specified different platform configuration for solution/service projects and sfproj build?

**Comment by ravipal** [comment]
The package targets (DotNetPublish and Publish) on service project isn't using the platform passed in from the pipeline task, it uses the default platform defined for the project which is 'Any CPU'. So before the package target runs, the service project should have been built using the default platform defined for the service project. For @jackbond case, the solution build is building the service projects using platform x64. My suggestion is to modify the solution configuration to build the service projects with platform 'Any CPU' even if the solution platform is x64 as shown below. No changes needed in pipeline task.
![image](https://user-images.githubusercontent.com/1775825/60985613-9d847600-a2f2-11e9-92db-6770cc1a48e9.png)

**Comment by jackbond** [comment]
@bishal-pdMSFT I changed the SF application project platform to AnyCPU and was able to build and deploy from within the pipeline. Thanks for the work around.

As to the root problem, I believe it would be better if the hosted agent pools didn't have service fabric installed at all by default. To avoid conflicts, it seems that the user should either installed the SDK via a Powershell task, or even better, Microsoft provided a build in task for installing the SDK (just like you do with .NET Core and Nuget.)

**Comment by bishal-pdMSFT** [comment]
@jackbond glad to know you are unblocked. Microsoft can not provide a task for installing SDK as SDK alters the global state of machine - something which is not desired.

@ravipal do you mean to say that the `Platform` input specified in "Build .sfproj" task is redundant and always 'AnyCPU' is used?
![image](https://user-images.githubusercontent.com/7830141/61354119-0f2e6800-a88f-11e9-9a50-074c662b3591.png)
Also, should the default pipeline be changed to use 'Any CPU' platform for solution build as well?

**Comment by athinanthny** [comment]
@jackbond  Please let us know if the issue still persists or is it resolved?

**Comment by maburlik** [comment]
Closing since the thread has not seen a reply in a while. Feel free to create another thread if you are still experiencing issues.
